README.md
Porting Advisor for Graviton
=============================

This is a fork of [Porting advisor](https://github.com/arm-hpc/porting-advisor), an open source project by the ARM High Performance Computing group. Originally, it was coded as a Python module that analyzed some known incompatibilities for C and Fortran code.

 It is a command line tool that analyzes source code for known code patterns and dependency libraries. It then generates a report with any incompatibilities with our Graviton processors. This tool provides suggestions of minimal required and/or recommended versions to run on Graviton instances for both language runtime and dependency libraries. It can run on non-ARM based machines (no Graviton processor needed). This tool does not work on binaries, just source code. It does not make any code modifications, it doesn’t make API level recommendations, nor does it send data back to AWS.

 This tool scans all files in a source tree, regardless of whether they are included by the build system or not. As such it may erroneously report issues in files that appear in the source tree but are excluded by the build system. Currently, the tool supports the following languages/dependencies:

* Python 3+
    * Python version
    * PIP version
    * Dependency versions in requirements.txt file
* Java 8+
    * Java version
    * Dependency versions in pom.xml file
    * JAR scanning for native method  calls (requires JAVA to be installed)
* Go 1.11+
    * Go version
    * Dependency versions on go.mod file
* C, C++, Fortran
    * Inline assembly with no corresponding aarch64 inline assembly.
    * Assembly source files with no corresponding aarch64 assembly source files.
    * Missing aarch64 architecture detection in autoconf config.guess scripts.
    * Linking against libraries that are not available on the aarch64 architecture.
    * Use of architecture specific intrinsic.
    * Preprocessor errors that trigger when compiling on aarch64.
    * Use of old Visual C++ runtime (Windows specific).
    * The following types of issues are detected, but not reported by default:
        * Compiler specific code guarded by compiler specific pre-defined macros.
    * The following types of cross-compile specific issues are detected, but not reported by default.
        * Architecture detection that depends on the host rather than the target.
        * Use of build artifacts in the build process.


For more information on how to modify issues reported, use the tool’s built-in help:

```
./porting-advisor-linux-x86_64 -–help
```

If you run into any issues, see our [CONTRIBUTING](CONTRIBUTING.md#reporting-bugsfeature-requests) file.

# How to run:

## As a Python script

**Pre-requisites**

- Python 3.10 or above (with PIP3 and venv module installed).
- (Optionally) Open JDK 17 (or above) and Maven 3.5 (or above) if you want to scan JAR files for native methods.

**Enable Python Environment**

Linux/Mac:
```
$. python3 -m venv .venv
$. source .venv/bin/activate
```

Powershell:
```
PS> python -m venv .venv
PS> .\.venv\Scripts\Activate.ps1
```

**Install requirements**
```
$. pip3 install -r requirements.txt
```

**Run tool (console output)**
```
$. python3 src/porting-advisor.py ~/my/path/to/my/repo
```

**Run tool (HTML report)**
```
$. python3 src/porting-advisor.py ~/my/path/to/my/repo --output report.html
```

## As a binary

### Generating the binary

**Pre-requisites**

- Python 3.10 or above (with PIP3 and venv module installed).
- (Optionally) Open JDK 17 (or above) and Maven 3.5 (or above) if you want the binary to be able to scan JAR files for native methods.

The `build.sh` script will generate a self-contained binary (for Linux/MacOS). It will be output to a folder called `dist`.

```
$ ./build.sh
```

For Windows, the `Build.ps1` will generate a folder with an EXE and all the files it requires to run.

```
PS> .\Build.ps1
```

**Running the binary**

**Pre-requisites**

Once you have the binary generated, it will only require Java 11 Runtime (or above) if you want to scan JAR files for native methods. Otherwise, the file is self-contained and doesn't need Python to run.

Default behaviour, console output:
```
$ ./porting-advisor-linux-x86_64 ~/my/path/to/my/repo
```

Generating HTML report:
```
$ ./porting-advisor-linux-x86_64 ~/my/path/to/my/repo --output report.html
```

Generating a report of just dependencies (this creates an Excel file with just the dependencies we found on the repo, no suggestions provided):
```
$ ./porting-advisor-linux-x86_64 ~/my/path/to/my/repo --output dependencies.xlsx --output-format dependencies
```

### Sample console report output:

```
./dist/porting-advisor-linux-x86_64 ./sample-projects/
| Elapsed Time: 0:00:03

Porting Advisor for Graviton v1.0.0
Report date: 2023-01-06 23:48:20

13 files scanned.
detected java code. we recommend using Corretto. see https://aws.amazon.com/corretto/ for more details.
detected python code. if you need pip, version 19.3 or above is recommended. we detected that you have version 22.2.1.
detected python code. min version 3.7.5 is required. we detected that you have version 3.10.6. see https://github.com/aws/aws-graviton-getting-started/blob/main/python.md for more details.
./sample-projects/java-samples/pom.xml: dependency library: leveldbjni-all is not supported on Graviton
./sample-projects/java-samples/pom.xml: using dependency library snappy-java version 1.1.3. upgrade to at least version 1.1.4
./sample-projects/java-samples/pom.xml: using dependency library zstd-jni version 1.1.0. upgrade to at least version 1.2.0
./sample-projects/python-samples/incompatible/requirements.txt:3: using dependency library OpenBLAS version 0.3.16. upgrade to at least version 0.3.17
detected go code. min version 1.16 is required. version 1.18 or above is recommended. we detected that you have version 1.15. see https://github.com/aws/aws-graviton-getting-started/blob/main/golang.md for more details.
./sample-projects/java-samples/pom.xml: using dependency library hadoop-lzo. this library requires a manual build  more info at: https://github.com/aws/aws-graviton-getting-started/blob/main/java.md#building-multi-arch-jars
./sample-projects/python-samples/incompatible/requirements.txt:5: dependency library NumPy is present. min version 1.19.0 is required.
detected java code. min version 8 is required. version 11 or above is recommended. see https://github.com/aws/aws-graviton-getting-started/blob/main/java.md for more details.

Use --output FILENAME.html to generate an HTML report.
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added 

- 

### Fixed

- 

## [1.0.0] - 2023-01-12

### Added

- AWS Graviton-Ready Assessor for Java by [Michael Fischer](https://github.com/otterley).
- Dependency scanner for Python - PIP.
- Added recommendation rules for Python.
- Dependency scanner for Java - Maven.
- Added recommendation rules for Java.
- Dependency scanner for Go.
- Added recommendation rules for Go.
- Dependency scanner for Node - NPM.
- Dependency scanner for C# - NuGet.
- Dependency scanner for Ruby.
- Added scripts to generate an executable binary.

### Changed

- HTML report formatting.

### Removed

- setup.py file as it required SUDO permissions.

Integration-Test.ps1
function Test-Line {
    param([string]$ReportType, [string]$Result, [string]$Pattern)

    if(-Not ($Result | Select-String -Quiet -Pattern $Pattern))
    {
        throw "**FAILED**: $ReportType report is missing: $Pattern"
    }
    else
    {
        Write-Host "**PASS** $ReportType report has: $Pattern"
    }
}

function Test-Report {
    param([string]$ReportType, [string]$Result, [string[]]$Patterns)

    ForEach($Pattern in $Patterns)
    {
        Test-Line $ReportType $Result $Pattern
    }
}

$Filename = .\Get-Binary-Name.ps1

$Lines_To_Find = @(
    "detected java code. we recommend using Corretto"
    "detected python code. min version 3.7.5 is required"
    "detected python code. if you need pip, version 19.3 or above is recommended"
    "using dependency library openblas version 0.3.16. upgrade to at least version 0.3.17"
    "dependency library numpy is present. min version 1.19.0 is required"
    "detected java code. min version 8 is required. version 11 or above is recommended"
    "using dependency library snappy-java version 1.1.3. upgrade to at least version 1.1.4"
    "using dependency library hadoop-lzo. this library requires a manual build"
    "dependency library: leveldbjni-all is not supported on Graviton"
    "detected go code. min version 1.16 is required. version 1.18 or above is recommended"
    "using dependency library github.com/golang/snappy version 0.0.1. upgrade to at least version 0.0.2"
)

Write-Host "Running samples to console"
$ResultConsole = Invoke-Expression ".\dist\$Filename\$Filename.exe .\sample-projects"
Test-Report "Console" $ResultConsole $Lines_To_Find

Write-Host "Running samples to HTML report"
Invoke-Expression ".\dist\$Filename\$Filename.exe .\sample-projects --output test.html"
$ResultHtml = Get-Content -Path test.html
Test-Report "HTML" $ResultHtml $Lines_To_Find
Remove-Item -Path test.html

Write-Host "Running samples to Dependency Report"
$Dependencies = @(
    "<si><t>component</t></si><si><t>version</t></si><si><t>origin</t></si><si><t>filename</t></si>"
    "<si><t>junit</t></si><si><t>4.8.2</t></si>"
    "<si><t>zstd-jni</t></si><si><t>1.1.0</t></si>"
    "<si><t>snappy-java</t></si><si><t>1.1.3</t></si>"
    "<si><t>lz4-java</t></si><si><t>1.4.0</t></si>"
    "<si><t>hadoop-lzo</t></si><si><t>0.4.17</t></si>"
    "<si><t>leveldbjni-all</t></si><si><t>1.8</t></si>"
    "<si><t>CommandLineParser</t></si><si><t>2.8.0</t></si>"
    "<si><t>Microsoft.Build.Utilities.Core</t></si><si><t>17.1.0</t></si>"
    "<si><t>Microsoft.Extensions.Logging.Console</t></si><si><t>6.0.0</t></si>"
    "<si><t>Microsoft.NET.Test.Sdk</t></si><si><t>16.5.0</t></si>"
    "<si><t>Microsoft.VisualStudio.Setup.Configuration.Interop</t></si><si><t>3.1.2196</t></si>"
    "<si><t>System.Linq.Async</t></si><si><t>6.0.1</t></si>"
    "<si><t>xunit</t></si><si><t>2.4.1</t></si>"
    "<si><t>coverlet.collector</t></si><si><t>1.2.0</t></si>"
    "<si><t>OpenBLAS</t></si>"
    "<si><t>SciPy</t></si>"
    "<si><t>NumPy</t></si>"
    "<si><t>FakeDependency</t></si><si><t>1.2.3</t></si>"
    "<si><t>0.3.17</t></si>"
    "<si><t>cors</t></si><si><t>2.8.5</t></si>"
    "<si><t>express</t></si><si><t>4.18.1</t></si>"
    "<si><t>rxjs</t></si><si><t>7.5.6</t></si>"
    "<si><t>socket.io</t></si><si><t>4.5.1</t></si>"
    "<si><t>@codechecks/client</t></si><si><t>0.1.12</t></si>"
    "<si><t>@commitlint/cli</t></si><si><t>17.0.3</t></si>"
    "<si><t>eslint</t></si><si><t>7.32.0</t></si>"
    "<si><t>typescript</t></si><si><t>4.7.4</t></si>"
    "<si><t>github.com/aws/aws-sdk-go</t></si>"
    "<si><t>github.com/golang/snappy</t></si>"
    "<si><t>rails</t></si><si><t>6.1.6.1</t></si>"
    "<si><t>rake</t></si><si><t>11.1</t></si>"
    "<si><t>actionpack</t></si>"
    "<si><t>bcrypt</t></si><si><t>3.1</t></si>"
    "<si><t>cucumber</t></si><si><t>4.1</t></si>"
    "<si><t>gc_tracer</t></si>"
    "<si><t>gssapi</t></si>"
    "<si><t>mail</t></si>"
    "<si><t>turbo-rails</t></si>"
    "<si><t>httpclient</t></si>"
    "<si><t>jruby-openssl</t></si>"
)
Invoke-Expression ".\dist\$Filename\$Filename.exe .\sample-projects --output test.xlsx --output-format dependencies"
# xlsx files are compressed files, so we need to unzip them and then compare them
Expand-Archive test.xlsx -DestinationPath temp
$ResultXlsx = Get-Content ".\temp\xl\sharedStrings.xml"
Test-Report "Dependencies" $ResultXlsx $Dependencies
Remove-Item -Path test.xlsx
Remove-Item -LiteralPath ".\temp" -Force -Recurse


Write-Host "--- Running negative tests ---"

Write-Host "Running missing arguments test"
$MissingArgumentError = $( $Result = & Invoke-Expression ".\dist\$Filename\$Filename.exe" ) 2>&1
if(-Not ($MissingArgumentError | Select-String -Quiet -Pattern "porting-advisor: error: the following arguments are required: DIRECTORY"))
{
    throw "**FAILED**: missing arguments test"
}
else
{
    Write-Host "**PASSED** missing arguments test"
}


Write-Host "Running directory not found test"
$DirectoryNotFoundError = $( $Result = & Invoke-Expression ".\dist\$Filename\$Filename.exe unexisting_directory" ) 2>&1
if(-Not ($DirectoryNotFoundError | Select-String -Quiet -Pattern "unexisting_directory: directory not found."))
{
    throw "**FAILED**: directory not found test"
}
else
{
    Write-Host "**PASSED** directory not found test"
}

exit 0